src/boot/lib/ast.ml

```diff
@@ -478,6 +480,139 @@ let ty_info = function
   | TyApp (fi, _, _) ->
       fi
 
+let const_has_side_effect = function
+  | CBool _
+  | CInt _
+  | Caddi _
+  | Csubi _
+  | Cmuli _
+  | Cdivi _
+  | Cmodi _
+  | Cnegi
+  | Clti _
+  | Cleqi _
+  | Cgti _
+  | Cgeqi _
+  | Ceqi _
+  | Cneqi _
+  | Cslli _
+  | Csrli _
+  | Csrai _
+  | Carity ->
+      false
+  (* MCore intrinsics: Floating-point numbers *)
+  | CFloat _
+  | Caddf _
+  | Csubf _
+  | Cmulf _
+  | Cdivf _
+  | Cnegf
+  | Cltf _
+  | Cleqf _
+  | Cgtf _
+  | Cgeqf _
+  | Ceqf _
+  | Cneqf _
+  | Cfloorfi
+  | Cceilfi
+  | Croundfi
+  | Cint2float
+  | Cstring2float ->
+      false
+  (* MCore intrinsics: Characters *)
+  | CChar _ | Ceqc _ | Cchar2int | Cint2char ->
+      false
+  (* MCore intrinsic: sequences *)
+  | Ccreate _
+  | Clength
+  | Cconcat _
+  | Cget _
+  | Cset _
+  | Ccons _
+  | Csnoc _
+  | CsplitAt _
+  | Creverse
+  | Csubsequence _ ->
+      false
+  (* MCore intrinsics: Random numbers *)
+  | CrandIntU _ ->
+      true
+  | CrandSetSeed ->
+      true
+  (* MCore intrinsics: Time *)
+  | CwallTimeMs ->
+      true
+  | CsleepMs ->
+      true
+  (* MCore intrinsics: Debug and I/O *)
+  | Cprint
+  | Cdprint
+  | CreadLine
+  | CreadBytesAsString
+  | CreadFile
+  | CwriteFile _
+  | CfileExists
+  | CdeleteFile
+  | Cerror
+  | Cexit ->
+      true
+  (* MCore intrinsics: Symbols *)
+  | CSymb _ | Cgensym | Ceqsym _ | Csym2hash ->
+      true
+  (* MCore intrinsics: References *)
+  | Cref | CmodRef _ | CdeRef ->
+      true
+  (* MCore intrinsics: Maps *)
+  (* NOTE(Linnea, 2021-01-27): Obj.t denotes the type of the internal map (I was so far unable to express it properly) *)
+  | CMap _
+  | CmapEmpty
+  | CmapSize
+  | CmapGetCmpFun
+  | CmapInsert _
+  | CmapRemove _
+  | CmapFindWithExn _
+  | CmapFindOrElse _
+  | CmapFindApplyOrElse _
+  | CmapMem _
+  | CmapAny _
+  | CmapMap _
+  | CmapMapWithKey _
+  | CmapFoldWithKey _
+  | CmapBindings
+  | CmapEq _
+  | CmapCmp _ ->
+      false
+  (* MCore intrinsics: Tensors *)
+  | CtensorCreate _
+  | CtensorGetExn _
+  | CtensorSetExn _
+  | CtensorRank
+  | CtensorShape
+  | CtensorCopyExn _
+  | CtensorReshapeExn _
+  | CtensorSliceExn _
+  | CtensorSubExn _
+  | CtensorIteri _ ->
+      true
```

Comment:
I don't think all the tensor intrinsics have side-effects. For example `tensorRank` does not modify the tensor.

Reply:
That is true, the same applies to `tensorCreate`, `tensorGetExn`, `tensorShape`, `tensorReshapeExn`, `tensorSliceExn`, and `tensorSubExn`. The only ones with side effects are those that return `unit`.

Reply:
I am waiting with that, for the same reasons as for symbols.